lib/galaxy/tools/__init__.py

```diff
@@ -1822,6 +1822,7 @@ def to_json(self, trans, kwd={}, job=None, workflow_building_mode=False):
         if self.input_translator:
             self.input_translator.translate(params)
 
+        set_dataset_matcher_factory(request_context, self)
```

Comment:
So the performance critical part of this commit is to setup the dataset_matcher earlier so that `populate_state `can benefit from the optimizations and the rest is (nice!) refactoring, right?

Reply:
I needed to drop a the input values parameter to move this earlier - so that looks like a refactoring but it was definitely part of this performance fix. The part that is really just pure refactoring is switching from comparing type strings to using the “type” attribute that I had forgotten I added to input types.